lib/services/downloads_service_backend.dart

```diff
@@ -0,0 +1,1515 @@
+      unawaited(_advanceQueue());
+      await _callbacksComplete!.future;
+      _enqueueLog.info("All downloads enqueued.");
+    } finally {
+      _callbacksComplete = null;
+    }
+  }
+
+  /// Advance the queue if possible and ready, no-op if not.
+  /// Will loop until all downloads have been enqueued.  Will enqueue
+  /// finampSettings.maxConcurrentDownloads at once.
+  Future<void> _advanceQueue() async {
+    try {
+      while (true) {
+        var nextTasks = _isar.downloadItems
+            .where()
+            .stateEqualTo(DownloadItemState.enqueued)
+            .filter()
+            .allOf(_activeDownloads,
+                (q, element) => q.not().isarIdEqualTo(element))
+            .limit(20)
+            .findAllSync();
+        if (nextTasks.isEmpty ||
+            !_downloadsService.allowDownloads ||
+            FinampSettingsHelper.finampSettings.isOffline) {
+          return;
+        }
+        for (var task in nextTasks) {
+          if (task.file == null) {
+            _enqueueLog
+                .severe("Recieved ${task.name} with no valid file path.");
+            _isar.writeTxnSync(() {
+              _downloadsService.updateItemState(task, DownloadItemState.failed);
+            });
+            continue;
+          }
+          // Do not limit enqueued downloads on IOS, it throttles them like crazy on its own.
+          while (_activeDownloads.length >=
+                      FinampSettingsHelper
+                          .finampSettings.maxConcurrentDownloads &&
+                  !Platform.isIOS ||
+              _finampUserHelper.currentUser == null) {
+            await Future.delayed(const Duration(milliseconds: 500));
+          }
+          await SchedulerBinding.instance.scheduleTask(() {
+            _activeDownloads.add(task.isarId);
+            // Base URL shouldn't be null at this point (user has to be logged in
+            // to get to the point where they can add downloads).
+            var url = switch (task.type) {
+              DownloadItemType.song => _jellyfinApiData
+                  .getSongDownloadUrl(
+                      item: task.baseItem!,
+                      transcodingProfile: task.fileTranscodingProfile)
+                  .toString(),
+              DownloadItemType.image => _jellyfinApiData
+                  .getImageUrl(
+                    item: task.baseItem!,
+                    // Download original file
+                    quality: null,
+                    format: null,
+                  )
+                  .toString(),
+              _ => throw StateError("???"),
+            };
+            _enqueueLog.fine(
+                "Submitting download ${task.name} to background_downloader.");
+            var downloadTask = DownloadTask(
+                taskId: task.isarId.toString(),
+                url: url,
+                //requiresWiFi:
+                //    FinampSettingsHelper.finampSettings.requireWifiForDownloads,
```

Comment:
I guess this isn't needed anymore with the new feature of background_downloader?

Reply:
Nope.